aws/session: Add support for shared config (~/.aws/config) file

jasdel opened the pull request:
Updates the SDK to take advantage of the new `AWS_SDK_LOAD_CONFIG` environment variable. When set to a truthy value, as defined in strconv.ParseBool, the SDK will load both shared config file (~/.aws/config), and the shared credentials file (~/.aws/credentials). The shared credentials file has precedence over the shared config.

Both config files share the same format. Both `[name]` and `[profile name]` are accepted as section names. The first, `[name]` will be looked for first before falling back to `[profile name]` per config file.

See the [session](http://docs.aws.amazon.com/sdk-for-go/api/aws/session/) package docs for more information.

Deprecates the `session.New` method since it cannot return an error when the session fails to load until a request is made. `session.NewSession` corrects this by returning an error when creating the session.

Fix #384

xibz commented on aws/session/env_config.go:
```diff
@@ -0,0 +1,188 @@
+	return cfg
+}
+
+func setEnvValue(dst *string, keys []string) {
```
This was a little confusing to me by the name. I was thinking this was exporting to some environment variable. Perhaps `getEnvValue(key []string) string` may make more sense or something along those lines?

jasdel replied:
Good point that makes a lot more sense. I'll update the naming.

xibz commented on aws/session/shared_config.go:
```diff
@@ -0,0 +1,120 @@
+// the profile provided. A sharedConfig pointer type value is used so that
+// multiple shared config file loadings can be chained.
+//
+// Only loads complete logically grouped values, and will no set fields in cfg
```
*will not set fields

xibz commented on aws/session/env_config.go:
```diff
@@ -0,0 +1,188 @@
+
+	// Require logical grouping of credentials
+	if len(cfg.Creds.AccessKeyID) == 0 || len(cfg.Creds.SecretAccessKey) == 0 {
+		cfg.Creds = credentials.Value{}
```
What is the provider name here? Do we care about empty `ProviderName`'s?

xibz commented on aws/session/env_config.go:
```diff
@@ -0,0 +1,188 @@
+// the shared SDK config will be loaded in addition to the SDK's specific
+// configuration values.
+func loadEnvConfig() envConfig {
+	enableSharedConfig, _ := strconv.ParseBool(os.Getenv("AWS_SDK_LOAD_CONFIG"))
```
Do we need a separate boolean here? Couldn't we just check `AWS_SHARED_CREDENTIALS_FILE`'s length to be greater than 0?

xibz commented:
Looks good with the few comments I have.

dlsniper commented on aws/session/session.go:
```diff
@@ -21,36 +15,128 @@ import (
+	return buildSession(envCfg, cfgs...)
+}
+
+func oldNewSession(cfgs ...*aws.Config) *Session {
```
I suggest renaming this to something more obvious like ` newSessionFallback ` or similar.

jasdel replied:
Thanks I'll take a look at refactoring these methods to see if I can create a better naming scheme for them, that fits in.

dlsniper commented on aws/session/session.go:
```diff
@@ -21,36 +15,128 @@ import (
+// Session and configuration at startup. Such as:
+//
+//     var sess = session.Must(session.Build())
+func Must(sess *Session, err error) *Session {
```
I would not introduce this at all, if the client consumes the result of `session.Build()` then it should handle the error accordingly, be it with ` panic() ` or other means.

jasdel replied:
That sounds fine, can remove this method. It has a nitch usage and can add it back later if there is demand for it.

dlsniper commented on aws/session/session.go:
```diff
@@ -21,36 +15,128 @@ import (
+		s, err := buildSession(envCfg, cfgs...)
+		if err != nil {
+			// TODO error message
+			panic(err)
```
Since the previous behavior is not a panic, at least as far as I can see, I wouldn't introduce it here either. Old behavior should be used until such time a 2.0 bc break version is done in which case this should be changed to return `*Session, error`.

jasdel replied:
Thanks, I agree, I'm not happy with the panic either. Logging the error at this point is the only alternative I can think of.

jasdel replied:
Updated the panic to follow the existing functionality of `New` where errors are not reported when request is made.

jasdel commented:
Fixed loading environment variable for the shared config file, and updated New error handling to not panic, and be inline with how errors would of been reported when requests are made.

jasdel commented on awstesting/integration/customizations/s3/s3manager/integration_test.go:
```diff
@@ -58,7 +58,7 @@ func setup() {
 // Delete the bucket
 func teardown() {
-	svc := s3.New(session.New())
+	svc := s3.New(session.NewFromSharedConfig())
```
Need to update integ tests for correct way to create session.

jasdel commented:
examples will be updated in another PR

xibz commented on aws/session/shared_config.go:
```diff
@@ -0,0 +1,214 @@
+	// Trim files from the list that don't exist.
+	for _, f := range files {
+		if err := cfg.setFromIniFile(profile, f); err != nil {
+			if aerr, ok := err.(awserr.Error); ok && aerr.Code() == "SharedConfigProfileNotExist" {
```
`SharedConfigProfileNotExist` probably should be a const, since used in multiple places

jasdel replied:
Sounds good was thinking about converting these to typed errors for simplicity anyways, so this would work out well.

xibz commented on aws/session/shared_config.go:
```diff
@@ -0,0 +1,214 @@
+
+func (cfg *sharedConfig) setFromIniFiles(profile string, files []sharedConfigFile) error {
+	// Trim files from the list that don't exist.
+	for _, f := range files {
```
this keeps iterating if an ini file has been found. Do we want to break if `err == nil`?

jasdel replied:
This is actually intended. The files are read in order and each file read has the chance to stomp over the previous file's values. This simplified the logic for reading/setting values, since we support some partial data loaded from each file. e.g credentials from first file, and region from second file.

xibz commented:
@jasdel - Looks pretty good! I have a few comments I've added.

xibz commented:
LGTM, :shipit:
